File: trees/BinarySearchTreeInsertion.cpp
/* Copyrights to venkys.io  */
/* For more information, visit https://venkys.io */
#include<bits/stdc++.h>

class Node {
public:
    int data;
    Node* left = NULL;
    Node* right = NULL;

    Node(int val) {
        data = val;
    }
};

// Function to insert a node into a Binary Search Tree (BST)
Node* insertBST(Node* root, int data) {
    if (root == NULL) return new Node(data);
    if (root->data == data) return root;
    else if (data < root->data) root->left = insertBST(root->left, data);
    else root->right = insertBST(root->right, data);
    return root;
}

// Function to perform an inorder traversal of the BST and print the elements
void inorder(Node* root) {
    if (root != NULL) {
        inorder(root->left);
        std::cout << root->data << " ";
        inorder(root->right);
    }
}

int main() {
    // User input for the number of elements in the BST
    std::cout << "Enter the number of elements in the BST: ";
    int n;
    std::cin >> n;

    Node* root = NULL;

    // User input for the elements of the BST
    std::cout << "Enter the elements of the BST:" << std::endl;
    for (int i = 0; i < n; i++) {
        int data;
        std::cin >> data;
        root = insertBST(root, data);
    }

    // Display the inorder traversal of the BST
    std::cout << "Inorder traversal of the BST: ";
    inorder(root);

    return 0;
}

File: trees/BinarySearchTreeInsertion.py
#  Copyrights to venkys.io
#  For more information, visit https://venkys.io

class Node:
    def __init__(self, data):
        # Constructor for the Node class
        self.data = data
        self.left = self.right = None

def insertBST(root, data):
    # Function to insert a node into the Binary Search Tree (BST)
    if root is None:
        # If the tree is empty, create a new node
        return Node(data)
    if root.data == data:
        # If the data is already present, return the current root
        return root
    elif data < root.data:
        # If the data is smaller, go to the left subtree
        root.left = insertBST(root.left, data)
    else:
        # If the data is larger, go to the right subtree
        root.right = insertBST(root.right, data)
    return root

def inorder(root):
    # Function to perform inorder traversal of the BST
    if root:
        inorder(root.left)
        print(root.data, end=" ")
        inorder(root.right)

if __name__ == "__main__":
    root = None
    
    # Taking user input for the BST elements
    user_input = input("Enter elements for the BST (separated by spaces): ")
    arr = list(map(int, user_input.split()))

    # Building the BST using user-provided elements
    for i in arr:
        root = insertBST(root, i)

    # Displaying the inorder traversal of the BST
    print("Inorder traversal of the BST:")
    inorder(root)
